File: kotlin/divide&conquer/main_1992.kt
lateinit var array: Array<IntArray>
var n = 0

fun main() {
    n = readLine()!!.toInt()
    array = Array(n) { readLine()!!.map { it - '0' }.toIntArray() }
    print(toQuad(n, 0 to 0))
}

fun toQuad(length: Int, start: Pair<Int, Int>): String {
    val standard = array[start.first][start.second]
    if (length == 1){
        return standard.toString()
    }

    var pass = true
    for (i in start.first until start.first + length) {
        for (j in start.second until start.second + length) {
            if (array[i][j] != standard) {
                pass = false
            }
        }
        if (!pass)
            break
    }
    return if (pass) {
        standard.toString()
    } else {
        val half = length / 2
        "(${toQuad(half, start)}${toQuad(half, start.first to start.second + half)}${toQuad(half, start.first + half to start.second)}${toQuad(half, start.first + half to start.second + half)})"
    }
}

File: cpp/divide&conquer/main_1992.cpp
#include <iostream>
#include <string>
using namespace std;

string* arr;

string cut(pair<int, int> start, int length) {
	const char standard = arr[start.first][start.second];
	if (length == 1) {
		if (standard == '0') {
			return "0";
		} else {
			return "1";
		}
	} else {
		bool pass = true;
		for (int i = start.first; i < start.first + length; ++i) {
			for (int j = start.second; j < start.second + length; ++j) {
				if (standard != arr[i][j]) {
					pass = false;
					break;
				}
			}
			if (!pass) {
				break;
			}
		}
		if (pass) {
			if (standard == '0') {
				return "0";
			} else {
				return "1";
			}
		} else {
			const int len = length / 2;
			string str = "(";
			str.append(cut(start, len));
			str.append(cut(make_pair(start.first, start.second + len), len));
			str.append(cut(make_pair(start.first + len, start.second), len));
			str.append(cut(make_pair(start.first + len, start.second + len), len));
			return str + ")";
		}
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	int n;
	cin >> n;
	arr = new string [n];
	for (int i = 0; i < n; ++i) {
		cin >> arr[i];
	}
	cout << cut(make_pair(0, 0), n);
}
